PurchasesCoreSwift/Attribution/AttributionFetcher.swift

```diff
@@ -0,0 +1,139 @@
+        // Should match available platforms in
+        // https://developer.apple.com/documentation/iad/adclient?language=swift
+        #if os(iOS)
+        guard let adClientProxy = attributionFactory.adClientProxy() else {
+            Logger.warn(Strings.attribution.search_ads_attribution_cancelled_missing_iad_framework)
+            completion(nil, AttributionFetcherError.identifierForAdvertiserFrameworksUnavailable)
+            return
+        }
+        adClientProxy.requestAttributionDetails(completion)
+        #else
+        completion(nil, AttributionFetcherError.identifierForAdvertiserUnavailableForPlatform)
+        #endif
+    }
+
+    @objc public var isAuthorizedToPostSearchAds: Bool {
+        // Should match platforms that require permissions detailed in
+        // https://developer.apple.com/app-store/user-privacy-and-data-use/
+        if !appTrackingTransparencyRequired {
+            return true
+        }
+
+        if #available(iOS 14.0.0, tvOS 14.0.0, *) {
+            return isAuthorizedToPostSearchAdsInATTRequiredOS
+        }
+
+        return true
+    }
+}
+
+private extension AttributionFetcher {
+
+    @available(iOS 14.0.0, tvOS 14.0.0, *)
+    private var isAuthorizedToPostSearchAdsInATTRequiredOS: Bool {
+        let minimumOSVersionRequiringAuthorization = OperatingSystemVersion(majorVersion: 14,
+                                                                            minorVersion: 5,
+                                                                            patchVersion: 0)
+        let needsTrackingAuthorization = systemInfo.isOperatingSystemAtLeastVersion(minimumOSVersionRequiringAuthorization)
+
+        guard let trackingManagerProxy = attributionFactory.atTrackingProxy() else {
+            if needsTrackingAuthorization {
+                Logger.warn(Strings.attribution.search_ads_attribution_cancelled_missing_att_framework)
+            }
+            return !needsTrackingAuthorization
+        }
+
+        let authStatusSelector = NSSelectorFromString(trackingManagerProxy.authorizationStatusPropertyName)
+        guard trackingManagerProxy.responds(to: authStatusSelector) else {
+            Logger.warn(Strings.attribution.att_framework_present_but_couldnt_call_tracking_authorization_status)
+            return false
+        }
+
+        // we use unsafeBitCast to prevent direct references to tracking frameworks, which cause issues for
+        // kids apps when going through app review, even if they don't actually use them at all.
+        typealias ClosureType = @convention(c) (AnyObject, Selector) -> FakeTrackingManagerAuthorizationStatus
+        let authStatusMethodImplementation = trackingManagerProxy.method(for: authStatusSelector)
+        let authStatusMethod: ClosureType = unsafeBitCast(authStatusMethodImplementation, to: ClosureType.self)
+        let authStatus = authStatusMethod(trackingManagerProxy, authStatusSelector)
```

Comment:
the objective-c [implementation uses `NSInvocation`](https://github.com/RevenueCat/purchases-ios/blob/main/Purchases/Attribution/RCAttributionFetcher.m#L105), but that's not available in Swift, so I'm using this instead. 

I tested it and it works correctly on device. 

As a bonus the type gets nicely casted to the `enum`, so the[ handling of the `authStatus`](https://github.com/RevenueCat/purchases-ios/blob/main/Purchases/Attribution/RCAttributionFetcher.m#L114) looks clearer now

Reply:
I've never had the opportunity to use `@convention(c)` 👏

Reply:
I'd never had the opportunity to google it before this 😂